src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingProcessor.cs

```diff
@@ -0,0 +1,418 @@
+
+        private async Task AcquirePartitions(CancellationToken ct)
+        {
+            // get all the pods in active status
+            var activePods = ListBlobs(_containerClient, "activepods");
+            var activePodCount = await GetActiveProcessors(activePods, _containerClient);
+            _totalProcessorsRunning = activePodCount;
+            Console.WriteLine($"Active Pod Count: {activePodCount}");
+
+            if (activePodCount == 0)
+            {
+                throw new Exception("active pods is 0");
+            }
+
+            // divide the partitions among the pods
+            // todo: handle cases where there is a remainder
+            // for example 8 partitions and 6 pods, which would require 4 pods to have 1 partition, and 2 pods to have 2 partitions
+            // for now, we will round down
+            var suggestedPartitionsPerPod = Math.Floor((double)_maximumProcessors / activePodCount);
+            Console.WriteLine($"Suggested Partitions Per Pod: {suggestedPartitionsPerPod}");
+
+            // find partitions without owners
+            // then aquire partitions one by one using ETags
+            //
+            // todo: detect if this takes too long
+            // for example, if we acquire 2 partitions right away but need 2 more and cannot get them,
+            // should we renew the 2 that we have, or not? Alternatively we could quit after X minutes
+            _ownedPartitions.Clear();
+
+            while (_ownedPartitions.Count < suggestedPartitionsPerPod)
+            {
+                HashSet<int> availablePartitions = new HashSet<int>();
+                for (int i = 0; i < _maximumPartitions; i++)
+                {
+                    availablePartitions.Add(i);
+                }
+
+                Dictionary<int, BlobItem> blobOwnerhip = new Dictionary<int, BlobItem>();
+
+                // get owned partitions
+                var ownershipList = ListBlobs(_containerClient, "ownedpartitions");
+
+                // remove owned partitions from set of available partitions
+                foreach (var ownership in ownershipList)
+                {
+                    ownership.Metadata.TryGetValue("owner", out var owner);
+                    ownership.Metadata.TryGetValue("owneduntil", out var ownedUntil);
+
+                    int partitionId = int.Parse(ownership.Name.Split('/')[1]);
+
+                    // if the partition is still owned then remove it from the list of available partitions
+                    if (DateTime.Parse(ownedUntil) > DateTime.UtcNow)
+                    {
+                        availablePartitions.Remove(partitionId);
+                    }
+
+                    blobOwnerhip.Add(partitionId, ownership);
+                }
+
+                // if not available partitions then return and wait
+                if (availablePartitions.Count == 0)
+                {
+                    Console.WriteLine("No partitions are available... Waiting");
```

Comment:
We would need to properly implement our logging.

Reply:
added logging